strawberryfields/parameters.py

```diff
@@ -19,335 +19,334 @@
+            return p
+
+        #p = p.evalf()  # TODO sympy 1.4 has some bugs which prevent using this for now, workaround below
+        if 0:
+            # build the substitution dict for binding free and measured params
+            d = {}
+            for k in p.atoms(MeasuredParameter, FreeParameter):
+                d[k] = k._eval_evalf(None)
+            p = p.evalf(subs=d)
+            # TODO evalf cannot handle substitutions with arbitrary objects even if p is atomic, hence substituting e.g. numpy arrays or TF classes causes sympy to raise an error
+            # TODO the float() conversion below prevents symbolic params from being passed through, maybe the backend should do float() conversion?
+            if not p.is_real:
+                return complex(p)
+            if p.is_integer:
+                return int(p)
+            return float(p)
+
+        # using lambdify we can also substitute np.ndarrays and tf.Tensors for the atoms
+        atoms = list(p.atoms(MeasuredParameter, FreeParameter))
+        func = sympy.lambdify(atoms, p, 'numpy')
+        vals = [k._eval_evalf(None) for k in atoms]
+        return func(*vals)
+
+    ret = list(map(do_evaluate, params))
+    if scalar:
+        return ret[0]
+    return ret
+
+
+def par_is_symbolic(p):
+    """Returns True iff p is a symbolic parameter instance.
+
+    An array is symbolic if any of its elements are.
     """
-    return tuple(p.x for p in params)
+    if is_object_array(p):
+        return any(par_is_symbolic(k) for k in p)
+    return isinstance(p, sympy.Basic)
 
 
-class Parameter():
-    """Represents a parameter passed to a :class:`strawberryfields.ops.Operation` subclass constructor.
+def par_regref_deps(p):
+    """RegRef dependencies of an Operation parameter.
 
-    The supported parameter types are Python and NumPy numeric types, NumPy arrays, :class:`.RegRef` instances,
-    :class:`.RegRefTransform` instances, and certain TensorFlow objects. RegRef instances are internally represented as
-    trivial RegRefTransforms.
+    Returns the RegRefs that the parameter depends on through the :class:`MeasuredParameter`
+    atoms it contains.
 
-    All but the RR and TensorFlow parameters represent an immediate numeric value that
-    will not change. RR parameters can only be evaluated after the corresponding register
-    subsystems have been measured. TF parameters can be evaluated whenever, but they depend on TF objects that
-    are evaluated using :meth:`tf.Session.run`.
+    Args:
+        p (Any): Operation parameter
+
+    Returns:
+        set[RegRef]: RegRefs the parameter depends on
+    """
+    ret = set()
+    if is_object_array(p):
+        # p is an object array, possibly containing symbols
+        for k in p:
+            ret.update(par_regref_deps(k))
+    elif isinstance(p, sympy.Basic):
+        # p is a Sympy expression, possibly containing measured parameters
+        for k in p.atoms(MeasuredParameter):
+            ret.add(k.regref)
```

Comment:
how come `add()` is used here, but `update` is used above?

Reply:
`add` adds a single element to the set, `update` adds all the elements of another set.

Reply:
We handle object arrays recursively.